### ISSUE TITLE
Significant performance difference depending on how I use the "filter" method

### ISSUE THREAD

**Issue by Nicolas-SB** [pola-rs/polars#18531]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
from pyiceberg.catalog import load_catalog
import tracemalloc
import time


def profile_query_execution(lazy_frame: pl.LazyFrame):
    print("Query plan:")
    print(lazy_frame.explain())
    start = time.time()
    tracemalloc.start()
    print(f"Shape of the result: {lazy_frame.collect().shape}")
    _, peak = tracemalloc.get_traced_memory()
    print(f"Peak memory usage: {peak/(1024*1024)} GB;")
    tracemalloc.stop()
    end = time.time()
    print(f"Elapsed time: {end-start} seconds")


iceberg_catalog = load_catalog("glue", **{"type": "glue", "s3.region": "eu-central-1"})
iceberg_table = iceberg_catalog.load_table(("my_database", "my_iceberg_table"))
lf = pl.scan_iceberg(iceberg_table)

lf_and = lf.filter(
    (pl.col("partition_column_1") == "partition_value_1")
    & (pl.col("partition_column_2") == "partition_value_2")
)
lf_comma = lf.filter(
    pl.col("partition_column_1") == "partition_value_1",
    pl.col("partition_column_2") == "partition_value_2",
)
lf_chained = lf.filter(pl.col("partition_column_1") == "partition_value_1").filter(
    pl.col("partition_column_2") == "partition_value_2"
)

print("\nlf_and")
profile_query_execution(lf_and)

print("\nlf_comma")
profile_query_execution(lf_comma)

print("\nlf_chained")
profile_query_execution(lf_chained)
```

### Log output

```shell
lf_and
Query plan:
PYTHON SCAN []
PROJECT */24 COLUMNS
SELECTION: [([(col("partition_column_1")) == (String(partition_value_1))]) & ([(col("partition_column_2")) == (String(partition_value_2))])]
Shape of the result: (469, 24)
Peak memory usage: 12.000402450561523 GB;
Elapsed time: 0.9519569873809814 seconds

lf_comma
Query plan:
PYTHON SCAN []
PROJECT */24 COLUMNS
SELECTION: [([(col("partition_column_1")) == (String(partition_value_1))]) & ([(col("partition_column_2")) == (String(partition_value_2))])]
Shape of the result: (469, 24)
Peak memory usage: 21.01994037628174 GB;
Elapsed time: 38.96175289154053 seconds

lf_chained
Query plan:
PYTHON SCAN []
PROJECT */24 COLUMNS
SELECTION: [([(col("partition_column_1")) == (String(partition_value_1))]) & ([(col("partition_column_2")) == (String(partition_value_2))])]
Shape of the result: (469, 24)
Peak memory usage: 11.866539001464844 GB;
Elapsed time: 0.8844459056854248 seconds
```

### Issue description

I just noticed that the memory usage and runtime of my queries goes up significantly if I pass multiple expressions to the filter method instead of creating one bigger expression or chaining multiple filter methods behind each other. 

The query plan for all three versions looks exactly the same and the logic evaluates to the same result, so I am confused why it behaves differently.

### Expected behavior

All three versions should behave exactly the same. 

### Installed versions

<details>
--------Version info---------
Polars:              1.6.0
Index type:          UInt32
Platform:            macOS-14.6.1-arm64-arm-64bit
Python:              3.12.5 | packaged by conda-forge | (main, Aug  8 2024, 18:32:50) [Clang 16.0.6 ]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2023.12.2
gevent               <not installed>
great_tables         <not installed>
matplotlib           3.9.2
nest_asyncio         <not installed>
numpy                1.26.4
openpyxl             <not installed>
pandas               2.2.2
pyarrow              17.0.0
pydantic             2.8.2
pyiceberg            0.7.1
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
</details>

**Comment by orlp** [comment]
That is very odd. Are the results consistently reproducible? We have a pass that transforms the `pl.all_horizontal` that the multi-expression `filter` generates into binary ANDs, and from the explain it does appear to be applied.

**Comment by deanm0000** [comment]
The predicates that make it to https://github.com/pola-rs/polars/blob/6ff1c70d25f108306ea53138b11eb262288b7251/py-polars/polars/io/iceberg.py#L179-L187 seem to be incomplete and iceberg isn't predicate pruning so polars is just getting more data.

**Comment by Nicolas-SB** [comment]
Yes, the results are consistently reproducible. Sorry that I was not able to make it better reproducible in the example, since I work with a Glue Data Catalog with bunch of data in it.

@deanm0000 I can confirm that the variable "predicate" in this function is None for the case where more data is loaded

**Comment by Nicolas-SB** [comment]
It does run, and it returns an Expr that prints to:

`[(col("partition_column_1")) == (String(partition_value_1))].all_horizontal([[(col("partition_column_2")) == (String(partition_value_2))]])`

which seems plausible to me.

The `_pyexr` attribute of this Expr is then passed to another "filter" Method of the self._ldf Object of this class, which is a `PyLazyFrame` imported from `polars.polars`. This does not seem to be a normal python class, because I can not find it anywhere in the python code of the library?
